pkg/controller/controller.go

```diff
@@ -151,45 +173,84 @@ func (c *Controller) handlePodObject(obj interface{}, verb string) {
 
 			c.observePodCreate(pvcKey)
 		}
-		c.enqueuePVC(pvc)
+		c.enqueueObject(pvc)
 		return
 	}
 }
 
-func (c *Controller) enqueuePVC(obj interface{}) {
-	var key string
-	var err error
-	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+func (c *Controller) enqueueObject(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
 		runtime.HandleError(err)
 		return
 	}
 	c.queue.AddRateLimited(key)
 }
 
+func waitForCacheSync(controllerName string, stopCh <-chan struct{}, c *Controller, v reflect.Value, controller interface{},
+	uploadController *UploadController) error {
+
+	glog.V(3).Infof("starting %s Run loop", controllerName)
+	glog.V(3).Infof("Waiting for %s informer caches to sync", controllerName)
+
+	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
+		return errors.New("Timeout waiting for pvc cache sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	if v.Type() == reflect.TypeOf(uploadController) {
+		if ok := cache.WaitForCacheSync(stopCh, controller.(*UploadController).servicesSynced); !ok {
+			return errors.New("Timeout waiting for service cache sync")
+		}
+	}
+	glog.V(3).Infof("%s cache has synced", controllerName)
+	return nil
+}
+
 //Run is being called from cdi controllers
-func (c *Controller) run(threadiness int, stopCh <-chan struct{}, controller interface{}) error { //*CloneContorler
+func (c *Controller) run(threadiness int, stopCh <-chan struct{}, controller interface{}) error {
 	defer func() {
+		runtime.HandleCrash()
 		c.queue.ShutDown()
 	}()
-	glog.V(3).Infoln("Starting cdi controller Run loop")
+
 	if threadiness < 1 {
 		return errors.Errorf("expected >0 threads, got %d", threadiness)
 	}
 
-	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
-		return errors.New("Timeout waiting for pvc cache sync")
-	}
-	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
-		return errors.New("Timeout waiting for pod cache sync")
+	v := reflect.ValueOf(controller)
```

Comment:
That's a cool way to do the type check, I usually run away from "reflect" but it's a good use for it here I think.